transformer_engine/pytorch/cpu_offload.py

```diff
@@ -0,0 +1,410 @@
+        ...         logging.info("On get", saved_state)
+        ...         tensor, = saved_state
+        ...         return tensor
+        ... 
+        >>> a = torch.ones(5, requires_grad=True)
+        >>> b = torch.ones(5, requires_grad=True) * 2
+        >>> with DummyHook():
+        ...     y = a * b
+        ... 
+        On save tensor([1., 1., 1., 1., 1.], requires_grad=True)
+        On save tensor([2., 2., 2., 2., 2.], grad_fn=<MulBackward0>)
+        >>> y.sum().backward()
+        On get (tensor([1., 1., 1., 1., 1.], requires_grad=True),)
+        On get (tensor([2., 2., 2., 2., 2.], grad_fn=<MulBackward0>),)
+
+    """
+
+    def __init__(self) -> None:
+        self.inside_context = False
+    
+    def __enter__(self):
+        self.inside_context = True
+        torch._C._autograd._push_saved_tensors_default_hooks(
+            self.on_save_for_backward, 
+            self.on_get_saved_tensor
+            )
+    
+    def __exit__(self, *args: Any):
+        self.inside_context = False
+        torch._C._autograd._pop_saved_tensors_default_hooks()
+    
+
+    def on_save_for_backward(self, tensor: torch.Tensor) -> Any:
+        raise NotImplementedError("`on_save_for_backward: Callable[[torch.Tensor], Any]`" 
+                                  "is not implemented in CpuOffloadHook class. Inherit "
+                                  "this class and implement your custom hooks")
+
+    def on_get_saved_tensor(self, saved_state: Any) -> torch.Tensor:
+        raise NotImplementedError("`on_get_saved_tensors: Callable[[Any], torch.Tensor]`" 
+                                  "is not implemented in CpuOffloadHook class. Inherit "
+                                  "this class and implement your custom hooks")
+
+class CpuOffloadHookWithOffloadHandler(CpuOffloadSavedTensorHook):
+    """Contex-manager that offloads/recovers tensors through an offload hander.
+    
+    The hook just offloads/recovers the tensor object to the handler through `tensor_push` and `tensor_pop` interface. 
+    How the offload-handler manages the offloading, recovering or prefetching timing is transparent to this hook. 
+    """
+    def __init__(self, offload_handler, handler_extra_kwargs={}, debug=False) -> None:
+        self.debug = debug
+        self.offload_handler = offload_handler
+        self.handler_extra_kwargs = handler_extra_kwargs
+        super().__init__()
+    
+    def on_save_for_backward(self, tensor: torch.Tensor) -> Any:
+        retrieve_identifier = self.offload_handler.tensor_push(
+            tensor,
+            **self.handler_extra_kwargs 
+        )
+        return retrieve_identifier
+    
+    def on_get_saved_tensor(self, retrieve_identifier: Any) -> torch.Tensor:
+        tensor = self.offload_handler.tensor_pop(
+            retrieve_identifier, 
+            **self.handler_extra_kwargs
+        )
+        return tensor
+
+class OffloadHandler:
+    """A base class for CPU offload-handler defining two methods."""
```

Comment:
nit: what 2 methods? I would probably just say "A base class for CPU offload-handler"